### PR TITLE
feat:add redis binding invoke of del&get

### PR DESCRIPTION
**Pull request by wxbty**
Signed-off-by: x-shadow-man <1494445739@qq.com>

# Description

Corrections for pr #2186

## Issue reference

We strive to have all PR being opened based on an issue, where the problem or feature have been discussed prior to implementation.

Please reference the issue this PR will close: #2068

## Checklist

Please make sure you've completed the relevant tasks for this PR, out of the following list:

* [ ] Code compiles correctly
* [ ] Created/updated tests
* [ ] Extended the documentation / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#2901

### CODE REVIEW

**Comment by wxbty** [comment]
Modified #2186 as per your suggestion  PTAL @ItalyPaleAle  thx

File: bindings/redis/redis.go [hunk 1]
```diff
@@ -66,21 +66,50 @@ func (r *Redis) Ping() error {
...
-		key := val
-		_, err := r.client.Do(ctx, "SET", key, req.Data).Result()
-		if err != nil {
-			return nil, err
+	if key, ok := req.Metadata["key"]; ok && key != "" {
+		switch req.Operation {
+		case bindings.DeleteOperation:
+			var arg []string
+			arg = append(arg, req.Metadata["key"]) //at least one key
+			index := 1
```
Comment (ItalyPaleAle) [inline bindings/redis/redis.go:82]: Why is `index` starting from 1, if you always do `index-1` below? Shouldn't it start from 0?

File: bindings/redis/redis.go [hunk 2]
```diff
@@ -66,21 +66,50 @@ func (r *Redis) Ping() error {
...
+			for argKey := range req.Metadata {
+				argIndex := fmt.Sprintf("%s%d", "key", index-1)
```
Comment (ItalyPaleAle) [inline bindings/redis/redis.go:84]: `argIndex := "key" + strconv.Itoa(index-1)`

**Review by ItalyPaleAle — CHANGES_REQUESTED** [review]
Thanks! This looks much better already, but please do address the comments

File: bindings/redis/redis.go [hunk 3]
```diff
@@ -66,21 +66,50 @@ func (r *Redis) Ping() error {
...
-		_, err := r.client.Do(ctx, "SET", key, req.Data).Result()
-		if err != nil {
-			return nil, err
+	if key, ok := req.Metadata["key"]; ok && key != "" {
+		switch req.Operation {
+		case bindings.DeleteOperation:
+			var arg []string
+			arg = append(arg, req.Metadata["key"]) //at least one key
+			index := 1
+			for argKey := range req.Metadata {
```
[the remark this answers is not in the document]
Comment (wxbty) [reply bindings/redis/redis.go:83]: According to the redis link  of the issue #2068,  redis get/set only supports a single operation, and del supports multiple operations, so I did the same. If this is inconsistent, I will modify it.

File: bindings/redis/redis.go (hunk 1 above)
Comment (wxbty) [reply bindings/redis/redis.go:82]: For consistency, the default [key] must exist. At the same time, in order to support multiple keys, the latter will start from key1, which means that the index also starts from 1. If it is changed to a single key, this problem does not exist

File: bindings/redis/redis.go [hunk 4]
```diff
@@ -66,21 +66,50 @@ func (r *Redis) Ping() error {
...
+			err := r.client.Del(ctx, arg...).Err()
+			if err != nil {
+				return nil, err
+			}
+		case bindings.GetOperation:
+			data, err := r.client.Get(ctx, key).Result()
+			if err != nil {
+				return nil, err
+			}
+			s, _ := strconv.Unquote(fmt.Sprintf("%q", data))
```
[the remark this answers is not in the document]
Comment (wxbty) [reply bindings/redis/redis.go:100]: I've seen the directGet method in state/redis/redis.go used like this, I'm doing this for consistency. I have found the difference now, the state uses r.client.Do, and returns interface{}, I use r.client.Get, which has already returned string, there is no need to deal with it, I will modify it

File: bindings/redis/redis.go (hunk 2 above)
Comment (wxbty) [reply bindings/redis/redis.go]: ok

File: bindings/redis/redis.go [hunk 5]
```diff
@@ -66,21 +64,41 @@ func (r *Redis) Ping() error {
...
+			rep := &bindings.InvokeResponse{}
+			rep.Data = []byte(data)
+			return rep, nil
+		case bindings.CreateOperation:
+			_, err := r.client.Do(ctx, "SET", key, req.Data).Result()
+			if err != nil {
+				return nil, err
+			}
+		default:
+			return nil, errors.Errorf("invalid operation type: %s. Expected %s, %s, or %s",
```
Comment (ItalyPaleAle) [inline bindings/redis/redis.go:96]: `return nil, fmt.Errorf("invalid operation type: %s", req.Operation)`

Users can get the list of supported operations by invoking the Features method

File: bindings/redis/redis.go [hunk 6]
```diff
@@ -66,21 +64,41 @@ func (r *Redis) Ping() error {
...
+				req.Operation, bindings.DeleteOperation, bindings.GetOperation, bindings.CreateOperation)
 		}
-
 		return nil, nil
 	}
-
-	return nil, errors.New("redis binding: missing key on write request metadata")
+	return nil, errors.New("redis binding: missing key on request metadata")
```
Comment (ItalyPaleAle) [inline bindings/redis/redis.go:101]: typo: `...in request metadata`

**Review by ItalyPaleAle — CHANGES_REQUESTED** [review]
Looks good! Please fix the linter issues and the last 2 small comments and we should be good to go :)

File: bindings/redis/redis.go (hunk 6 above)
Comment (wxbty) [reply bindings/redis/redis.go:101]: ok

File: bindings/redis/redis.go (hunk 5 above)
Comment (wxbty) [reply bindings/redis/redis.go:96]: ok, but this will be a bit inconsistent with /bind/mysql.go

**Review by ItalyPaleAle — APPROVED** [review]

**Comment by wxbty** [comment]
hi, @berndverst   Can this pr be merged?

**Comment by berndverst** [comment]
> hi, @berndverst Can this pr be merged?

Yes, looks good to me! Will merge.

**Comment by berndverst** [comment]
@x-shadow-man @wxbty please create a docs PR against the 1.10 to update the supported operations for this binding.

**Comment by berndverst** [comment]
> > hi, @berndverst Can this pr be merged?
> 
> Yes, looks good to me! Will merge.

I cannot merge, because I don't have permission to update your PR / branch! Can you give me permission so I can keep the PR in sync with Master?

**Comment by berndverst** [comment]
```
git push git@github.com:x-shadow-man/components-contrib.git --force
Enumerating objects: 11, done.
Counting objects: 100% (11/11), done.
Delta compression using up to 8 threads
Compressing objects: 100% (6/6), done.
Writing objects: 100% (6/6), 1.08 KiB | 1.08 MiB/s, done.
Total 6 (delta 4), reused 0 (delta 0), pack-reused 0
remote: Resolving deltas: 100% (4/4), completed with 4 local objects.
To github.com:x-shadow-man/components-contrib.git
 ! [remote rejected]   feat-redis-dg -> feat-redis-dg (permission denied)
error: failed to push some refs to 'github.com:x-shadow-man/components-contrib.git'
```

**Comment by wxbty** [comment]
> ```
> git push git@github.com:x-shadow-man/components-contrib.git --force
> Enumerating objects: 11, done.
> Counting objects: 100% (11/11), done.
> Delta compression using up to 8 threads
> Compressing objects: 100% (6/6), done.
> Writing objects: 100% (6/6), 1.08 KiB | 1.08 MiB/s, done.
> Total 6 (delta 4), reused 0 (delta 0), pack-reused 0
> remote: Resolving deltas: 100% (4/4), completed with 4 local objects.
> To github.com:x-shadow-man/components-contrib.git
>  ! [remote rejected]   feat-redis-dg -> feat-redis-dg (permission denied)
> error: failed to push some refs to 'github.com:x-shadow-man/components-contrib.git'
> ```

done

**Comment by wxbty** [comment]
> @x-shadow-man @wxbty please create a docs PR against the 1.10 to update the supported operations for this binding.

please see dapr/docs#2901

**Comment by wxbty** [comment]
> [@ItalyPaleAle](/ItalyPaleAle)
> ** ItalyPaleAle ** approved these changes

**Comment by wxbty** [comment]
@ItalyPaleAle   sorry，misoperation of “requested a review again”

**Review by ItalyPaleAle — APPROVED** [review]

**Pull request merged**